File: app/javascript/tailmix/runtime/interpreter.js
import { StateOperations } from './operations/state';
import { CollectionsOperations } from './operations/collections';
import { LogicalOperations } from './operations/logical';
import { InteropOperations } from './operations/interop';
import { HttpOperations } from './operations/http';
import { ArithmeticOperations } from './operations/arithmetic';
import { ValueOperations } from './operations/value';
import { ComparisonOperations } from './operations/comparison';
import { VariableOperations } from './operations/variables';
import { ElementOperations } from './operations/element';
import { HtmlOperations } from './operations/html';
import { ThisOperations } from './operations/this';
import { DomOperations } from './operations/dom';

const OPERATIONS = {
    ...StateOperations,
    ...CollectionsOperations,
    ...LogicalOperations,
    ...InteropOperations,
    ...HttpOperations,
    ...ArithmeticOperations,
    ...ValueOperations,
    ...ComparisonOperations,
    ...VariableOperations,
    ...ElementOperations,
    ...ThisOperations,
    ...HtmlOperations,
    ...DomOperations
};

export class Interpreter {
    constructor(component) {
        this.component = component;
        this._responseContext = null;
    }

    async run(expressions, context = {}) {
        const executionContext = { ...context, vars: {} };

        if (!Array.isArray(expressions)) {
            console.error("Tailmix Interpreter Error: `run` was called with a non-array value.", expressions);
            return;
        }

        for (const expr of expressions) {
            await this.eval(expr, executionContext);
        }
    }

    async eval(expression, context) {
        if (!Array.isArray(expression)) {
            return expression;
        }

        if (expression.length === 0) {
            return null;
        }

        const [op, ...args] = expression;

        if (op === 'item') {
            return args.reduce((obj, key) => obj?.[key], context.item);
        }

        if (op === 'response') {
            if (!this._responseContext) {
                console.warn("Tailmix: `:response` can only be used inside a `fetch` callback.");
                return null;
            }
            // The `response` operation ALWAYS reads from the temporary response context.
            return args.reduce((obj, key) => obj?.[key], this._responseContext);
        }

        const handler = OPERATIONS[op];
        if (!handler) {
            console.warn(`Tailmix: Unknown operation "${op}"`);
            return;
        }

        // Call the handler, passing the interpreter instance, args, and the current context.
        return await handler(this, args, context);
    }

    async evalBranch(branch, context) {
        if (!branch) return;
        for (const expr of branch) {
            await this.eval(expr, context);
        }
    }
}

File: lib/tailmix/definition/scripting/interpreter.rb
# frozen_string_literal: true

# Require all operation modules
Dir[File.join(__dir__, "operations", "*.rb")].each { |file| require file }

module Tailmix
  module Definition
    module Scripting
      class Interpreter
        OPERATIONS = {
          # State
          set: Operations::State.method(:set),
          toggle: Operations::State.method(:toggle),
          increment: Operations::State.method(:increment),
          # Collections
          each: Operations::Each.method(:each),
          array_push: Operations::Collections.method(:array_push),
          array_remove_at: Operations::Collections.method(:array_remove_at),
          array_update_at: Operations::Collections.method(:array_update_at),
          array_remove_where: Operations::Collections.method(:array_remove_where),
          array_update_where: Operations::Collections.method(:array_update_where),
          # Logical
          and: Operations::Logical.method(:and),
          or: Operations::Logical.method(:or),
          not: Operations::Logical.method(:not),
          if: Operations::Logical.method(:if),
          # Interop
          call: Operations::Interop.method(:call),
          # Http
          fetch: Operations::Http.method(:fetch),
          # Arithmetic
          add: Operations::Arithmetic.method(:add),
          subtract: Operations::Arithmetic.method(:subtract),
          mod: Operations::Arithmetic.method(:mod),
          # Comparison
          eq: Operations::Comparison.method(:eq),
          lt: Operations::Comparison.method(:lt),
          gt: Operations::Comparison.method(:gt),
          # Value
          state: Operations::Value.method(:state),
          item: Operations::Value.method(:item),
          now: Operations::Value.method(:now),
          concat: Operations::Value.method(:concat),
          log: Operations::Value.method(:log),
          response: Operations::Value.method(:response),
          # Variables
          let: Operations::Variables.method(:let),
          var: Operations::Variables.method(:var),

          # DOM (placeholders)
          dom_append: Operations::Dom.method(:dom_append),
          dom_prepend: Operations::Dom.method(:dom_prepend),
          dom_add_class: Operations::Dom.method(:dom_add_class),
          dom_remove_class: Operations::Dom.method(:dom_remove_class),
          dom_toggle_class: Operations::Dom.method(:dom_toggle_class),
          dom_replace: Operations::Dom.method(:dom_replace),
          dom_remove: Operations::Dom.method(:dom_remove),
          dom_set_attribute: Operations::Dom.method(:dom_set_attribute),
          dom_set_value: Operations::Dom.method(:dom_set_value),

          # HTML
          html_build: Operations::Html.method(:html_build),
          element_attrs: Operations::Html.method(:element_attrs),
        }.freeze

        attr_reader :context, :actions_definition, :server_context
        attr_accessor :response_context

        def self.eval_all(expressions, context, actions_definition = {}, server_context = {})
          new(context, actions_definition, server_context).tap do |interpreter|
            expressions.each { |expr| interpreter.eval(expr) }
          end.context
        end

        def initialize(context, actions_definition = {}, server_context = {})
          @context = context.dup
          @actions_definition = actions_definition
          @server_context = server_context
          @response_context = nil
          @local_variables = {}
        end

        def eval(expression)
          op, *args = expression
          handler = OPERATIONS[op]
          raise Error, "Unknown operation: #{op}" unless handler
          handler.call(self, args)
        end

        def eval_with_response_context(expressions, response_context)
          self.response_context = response_context
          self.class.eval_all(expressions, @context, @actions_definition, @server_context)
          self.response_context = nil
        end

        def eval_branch(branch)
          return nil unless branch
          # Pass the actions_definition to the sub-evaluation
          self.class.eval_all(branch, @context, @actions_definition)
        end
      end
    end
  end
end
